src/objects.c

```diff
@@ -557,32 +558,26 @@ CK_RV p11prov_obj_find(P11PROV_CTX *provctx, P11PROV_SESSION *session,
     if (ret != CKR_OK) {
         return ret;
     }
-    while (total < MAX_OBJS_IN_STORE) {
-        CK_OBJECT_HANDLE object[OBJS_IN_SEARCH];
-        ret = p11prov_FindObjects(provctx, sess, object, OBJS_IN_SEARCH,
-                                  &objcount);
+    do {
+        objcount = 0;
+        objects = OPENSSL_realloc(objects, (total + OBJS_PER_SEARCH)
+                                               * sizeof(CK_OBJECT_HANDLE));
+        if (objects == NULL) {
+            (void)p11prov_FindObjectsFinal(provctx, sess);
+            return CKR_HOST_MEMORY;
+        }
+        ret = p11prov_FindObjects(provctx, sess, &objects[total],
+                                  OBJS_PER_SEARCH, &objcount);
         if (ret != CKR_OK || objcount == 0) {
             result = ret;
             break;
         }
         total += objcount;
+    } while (objcount != 0 && total < MAX_OBJS_IN_STORE);
```

Comment:
the `objcount != 0` is already checked on line 571 so this condition is  not needed.

Reply:
fair